packages/node/src/index.ts

```diff
@@ -381,40 +382,48 @@ export const build: BuildV3 = async ({
     ? handler.substring(0, handler.length - 3)
     : handler;
 
+  const isMiddleware = config.middleware === true;
+
   // Will output an `EdgeFunction` for when `config.middleware = true`
   // (i.e. for root-level "middleware" file) or if source code contains:
   // `export const config = { runtime: 'experimental-edge' }`
-  let isEdgeFunction = false;
+  let isEdgeFunction = isMiddleware;
+
+  const project = new Project();
+  const staticConfig = getConfig(project, entrypointPath);
+  if (staticConfig?.runtime) {
+    if (!ALLOWED_RUNTIMES.includes(staticConfig.runtime)) {
+      throw new Error(
+        `Unsupported "runtime" property in \`config\`: ${JSON.stringify(
+          staticConfig.runtime
+        )} (must be one of: ${JSON.stringify(ALLOWED_RUNTIMES)})`
+      );
+    }
+    isEdgeFunction = staticConfig.runtime === 'experimental-edge';
+  }
+
+  // Add a `route` for Middleware
+  if (isMiddleware) {
+    if (!isEdgeFunction) {
+      // Root-level middleware file can not have `export const config = { runtime: 'nodejs' }`
+      throw new Error(
+        `Middleware file can not be a Node.js Serverless Function`
+      );
+    }
```

Comment:
**nit:** not related to this PR, but I wonder if the following would be simpler to understand, by not having `isEdgeFunction` and `isMiddleware` and inlining the usage as the conditionals are pretty straightforward:

```ts
const providedRuntime = AllowedRuntimes.parse(staticConfig.runtime); // pseudo code
if (config.middleware === true) {
  if (providedRuntime && providedRuntime !== 'experimental-edge') {
  ...
  }
}
```